### PR TITLE
Re-write MZ bin plugin. Related issue: #2436

### PR DESCRIPTION
**Pull request by Nodepad**
* Add loading of segments like sections
* Add proper loading of entry point
* Add loading of relocation entries (If I understood properly relocations right now doesn't work, so I can't say if I create relocations properly).
* Also add function btree_traverse for enumeration of data in tree in regular or reverse order. (Function is needed for plugin).

### CODE REVIEW

File: libr/util/btree.c [hunk 1]
```diff
@@ -47,6 +47,24 @@ R_API void *btree_search(struct btree_node *root, void *x, BTREE_CMP(cmp), int p
 	} return NULL;
 }
 
+R_API void *btree_traverse(struct btree_node *root, int reverse, void *context, BTREE_TRV(trv))
+{
+	if (root!=NULL) {
+		if (reverse)
+		{
```
Comment (radare) [inline libr/util/btree.c]: Wrong indent. Follow the same rules as the rest of r2

**Comment by radare** [comment]
Do not define variables in the middle of nothing. Do not check for nonzero when you have to check for >0. Always nullchk when following tainted pointers. Care about file size boundaries. This code is full of vulnerabilities, could you update the pullreq fixing them? Thanks

**Comment by Nodepad** [comment]
1. By wrong indent you mean position of opening brackets?
2. Ok, I'll check and fix.

**Comment by radare** [comment]
Yep.

Cool! Thanks! Im being a bit picky trying to merge only safe stuff. Thanks for that rework. About the relocs: they are currently imported in r2 as flags so you can hint the disassembly with those labels. Future plans include assigning an io write cache associated to every rbinobject so it can be overlapped on top of the underlying io.

**Comment by Nodepad** [comment]
Updated. 
Changed indentation from spaces to tabs. Limited line width to 78 characters. Added checks for file size boundaries and pointer checks.

**Comment by jvoisin** [comment]
Please use `const` every time you can.

File: libr/bin/p/bin_mz.c [hunk 2]
```diff
@@ -1,210 +1,208 @@
...
-
-	r_list_append (ret, ptr);
+	ret->free = free;
+	if (!(segments = r_bin_mz_get_segments (arch->o->bin_obj))){
+		r_list_free (ret);
+		return NULL;
+	}
+	for (i = 0; !segments[i].last; i++) {
+		if (!(ptr = R_NEW0 (RBinSection)))
+			break;
```
Comment (jvoisin) [inline libr/bin/p/bin_mz.c]: This is an error case; shouldn't it free everything and return `NULL`?

**Comment by jvoisin** [comment]
Cool PR :)

File: libr/bin/p/bin_mz.c (hunk 2 above)
Comment (Nodepad) [inline libr/bin/p/bin_mz.c]: I made it like in bin_pe.c. But I think you're right, should return NULL.

**Comment by Nodepad** [comment]
Updated. Add const modifiers and make style of function calls consistent.

**Comment by radare** [comment]
Merged. Thanks!

**Pull request closed without merging**